// yaml-datastore/dist/index.js
// src/idgen.ts
import prand from "pure-rand";
var seed = 79 + 83 + 72 + 87;
var rng = prand.xorshift128plus(seed);
var chars = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "A", "B", "C", "D", "E", "F"];
function uniformints(lBound, uBound, numInts) {
  let result = [];
  if (Number.isInteger(lBound) && Number.isInteger(uBound) && Number.isInteger(numInts) && uBound > lBound && numInts >= 1) {
    for (let i = 0; i < numInts; i++) {
      result.push(prand.unsafeUniformIntDistribution(lBound, uBound, rng));
    }
  }
  return result;
}
function resetRNG() {
  rng = prand.xorshift128plus(seed);
}
function generateIDs(numIDs, numSkip) {
  resetRNG();
  const idLen = 6;
  const lBound = 0;
  const uBound = 15;
  let result = [];
  if (Number.isInteger(numIDs) && Number.isInteger(numSkip) && numIDs > 0 && numSkip >= 0) {
    for (let i = 0; i < numSkip; i++) {
      uniformints(lBound, uBound, idLen);
    }
    for (let i = 0; i < numIDs; i++) {
      let indices = uniformints(lBound, uBound, idLen);
      let id = chars[indices[0]] + chars[indices[1]] + chars[indices[2]] + chars[indices[3]] + chars[indices[4]] + chars[indices[5]];
      result.push(id);
    }
  }
  return result;
}

// src/load.ts
import path2 from "path";
import fs2 from "fs";
import yaml2 from "js-yaml";

// src/utils.ts
import path from "path";
import fs from "fs";
import yaml from "js-yaml";
var doubleParenthesesRegEx = new RegExp(/\(\(.*\)\)/);
function complexStringKeyToFileName(keyName) {
  const regex = new RegExp("(?<!_|^)_(?!_|$)", "g");
  return keyName.replaceAll(regex, ".");
}
var ElementPathResult = class {
  _type;
  _data;
  _parentElementPath;
  _parentFilePath;
  _keyName;
  /**
   * Default constructor for ElementPathResult
   *
   * @param type elementPath type
   * @param data filepath to be read into memory, or simple element, or null (for invalid path)
   * @param parentElementPath elementPath of parent
   * @param parentFilePath file path to parent element
   * @param keyName property in object or index in list
   */
  constructor(type, data, parentElementPath, parentFilePath, keyName) {
    this._type = type;
    if (this._type === 0 /* invalid */) {
      this._data = null;
    } else {
      this._data = data;
    }
    this._parentElementPath = parentElementPath;
    this._parentFilePath = parentFilePath;
    this._keyName = keyName;
  }
  /** @returns elementPath type */
  get type() {
    return this._type;
  }
  /** @returns filepath to be read into memory, or simple element, or null (for invalid path) */
  get data() {
    return this._data;
  }
  /** @returns elementPath of parent */
  get parentElementPath() {
    return this._parentElementPath;
  }
  /** @returns file path to parent element*/
  get parentFilePath() {
    return this._parentFilePath;
  }
  /** @returns property in object or index in list */
  get keyName() {
    return this._keyName;
  }
  /** @returns boolean value identifying if parent is an element */
  get parentIsElement() {
    return this._parentFilePath.slice(-5) === ".yaml" && fs.existsSync(this._parentFilePath);
  }
};
function trimDoubleParentheses(aString) {
  return aString.slice(2, -2);
}
function getNextElementPath(elementPath) {
  let elementPathToCheck = elementPath;
  let firstDotIndex = elementPathToCheck.indexOf(".");
  let firstBracketIndex = elementPathToCheck.indexOf("[");
  let firstElementEntry = "";
  if (firstDotIndex === 0) {
    elementPathToCheck = elementPath.slice(1);
    firstDotIndex = elementPathToCheck.indexOf(".");
    firstBracketIndex = elementPathToCheck.indexOf("[");
  } else if (firstBracketIndex === 0) {
    elementPathToCheck = elementPath.slice(1);
    let firstClosedBracketIndex = elementPathToCheck.indexOf("]");
    if (firstClosedBracketIndex !== -1) {
      firstElementEntry = elementPathToCheck.substring(
        0,
        firstClosedBracketIndex
      );
    }
    return firstElementEntry;
  }
  if (firstDotIndex === -1 && firstBracketIndex !== -1) {
    firstElementEntry = elementPathToCheck.substring(0, firstBracketIndex);
  } else if (firstBracketIndex === -1 && firstDotIndex !== -1) {
    firstElementEntry = elementPathToCheck.substring(0, firstDotIndex);
  } else if (firstBracketIndex !== -1 && firstDotIndex !== -1) {
    firstElementEntry = elementPathToCheck.substring(
      0,
      Math.min(firstBracketIndex, firstDotIndex)
    );
  } else {
    firstElementEntry = elementPathToCheck;
  }
  return firstElementEntry;
}
function getElementPathInfo(workingDirectoryPath, elementPath) {
  let parentElementPath = "";
  let parentFilePath = path.join(workingDirectoryPath, "..", "_this.yaml");
  let keyName = "";
  if (elementPath === "") {
    let filepath = path.join(workingDirectoryPath, "_this.yaml");
    if (fs.existsSync(filepath)) {
      if (!fs.existsSync(parentFilePath)) {
        parentFilePath = path.join(workingDirectoryPath, "..");
      }
      return new ElementPathResult(
        1 /* empty */,
        filepath,
        "",
        parentFilePath,
        keyName
      );
    }
  } else if (!elementPath.includes(".") && !elementPath.includes("[") && !elementPath.includes("]")) {
    let filePath = path.join(workingDirectoryPath, elementPath, "_this.yaml");
    if (fs.existsSync(filePath)) {
      let parentFilePath3 = path.join(workingDirectoryPath, "_this.yaml");
      if (!fs.existsSync(parentFilePath3)) {
        parentFilePath3 = workingDirectoryPath;
      }
      keyName = elementPath;
      return new ElementPathResult(
        2 /* shortToObject */,
        filePath,
        "",
        parentFilePath3,
        keyName
      );
    }
    filePath = path.join(workingDirectoryPath, elementPath + ".yaml");
    if (fs.existsSync(filePath)) {
      return new ElementPathResult(
        3 /* shortToList */,
        filePath,
        "",
        "",
        keyName
      );
    }
    const parentFilePath2 = path.join(workingDirectoryPath, "_this.yaml");
    keyName = elementPath;
    if (fs.existsSync(parentFilePath2)) {
      const thisYamlContent = fs.readFileSync(parentFilePath2, "utf-8");
      const yamlAsJsObj = yaml.load(thisYamlContent);
      const rawData = yamlAsJsObj[elementPath];
      if (typeof rawData === "string") {
        if (doubleParenthesesRegEx.test(rawData)) {
          filePath = path.join(
            workingDirectoryPath,
            trimDoubleParentheses(rawData)
          );
          if (fs.existsSync(filePath)) {
            return new ElementPathResult(
              4 /* shortToComplexString */,
              filePath,
              "",
              parentFilePath2,
              elementPath
            );
          }
        }
      }
      return new ElementPathResult(
        5 /* shortToSimple */,
        rawData,
        "",
        parentFilePath2,
        elementPath
      );
    } else {
      return new ElementPathResult(0 /* invalid */, null, "", "", null);
    }
  } else {
    const firstElementEntry = getNextElementPath(elementPath);
    parentElementPath = firstElementEntry;
    if (firstElementEntry === "" || firstElementEntry.includes(".") || firstElementEntry.includes("[") || firstElementEntry.includes("]")) {
      return new ElementPathResult(0 /* invalid */, null, "", "", null);
    }
    const firstElementPathInfo = getElementPathInfo(
      workingDirectoryPath,
      firstElementEntry
    );
    parentFilePath = firstElementPathInfo.data;
    if (firstElementPathInfo.type !== 0 /* invalid */) {
      const firstElementContent = fs.readFileSync(
        firstElementPathInfo.data,
        "utf-8"
      );
      const firstElementAsJsObj = yaml.load(firstElementContent);
      let remainingElementPath = elementPath.replace(firstElementEntry, "");
      let remainingElementEntries = [];
      do {
        const nextElementPath = getNextElementPath(remainingElementPath);
        if (nextElementPath === "" || nextElementPath.includes(".") || nextElementPath.includes("[") || nextElementPath.includes("]")) {
          return new ElementPathResult(
            0 /* invalid */,
            null,
            "",
            "",
            null
          );
        }
        remainingElementEntries.push(nextElementPath);
        if (remainingElementPath[0] === ".") {
          remainingElementPath = remainingElementPath.replace(
            "." + nextElementPath,
            ""
          );
        } else if (remainingElementPath[0] === "[") {
          remainingElementPath = remainingElementPath.replace(
            "[" + nextElementPath + "]",
            ""
          );
        }
      } while (remainingElementPath !== "");
      if (remainingElementEntries.length > 1) {
        for (const elementEntry of remainingElementEntries.slice(0, -1)) {
          parentElementPath = parentElementPath + "[" + elementEntry + "]";
          parentFilePath = getElementPathInfo(
            workingDirectoryPath,
            parentElementPath
          ).data;
        }
      }
      keyName = remainingElementEntries.slice(-1)[0];
      let currentElementAsJsObj = firstElementAsJsObj;
      let filePath = path.dirname(firstElementPathInfo.data);
      for (let i = 0; i < remainingElementEntries.length; i++) {
        const elementPath2 = remainingElementEntries[i];
        if (!currentElementAsJsObj.hasOwnProperty(elementPath2)) {
          return new ElementPathResult(
            0 /* invalid */,
            null,
            "",
            "",
            null
          );
        }
        const rawData = currentElementAsJsObj[elementPath2];
        if (typeof rawData === "string") {
          if (doubleParenthesesRegEx.test(rawData)) {
            if (filePath.slice(-5) === ".yaml") {
              filePath = path.join(
                filePath.split("/").slice(0, -1).join("/"),
                trimDoubleParentheses(rawData)
              );
            } else {
              filePath = path.join(filePath, trimDoubleParentheses(rawData));
            }
            const currentElementContent = fs.readFileSync(filePath, "utf-8");
            currentElementAsJsObj = yaml.load(currentElementContent);
            if (i === remainingElementEntries.length - 1) {
              if (filePath.slice(-10) === "_this.yaml") {
                const result2 = new ElementPathResult(
                  6 /* hierarchicalToObject */,
                  filePath,
                  parentElementPath,
                  parentFilePath,
                  keyName
                );
                return result2;
              } else if (filePath.slice(-5) === ".yaml") {
                const result2 = new ElementPathResult(
                  7 /* hierarchicalToList */,
                  filePath,
                  parentElementPath,
                  parentFilePath,
                  keyName
                );
                return result2;
              } else {
                return new ElementPathResult(
                  8 /* hierarchicalToComplexString */,
                  filePath,
                  parentElementPath,
                  parentFilePath,
                  keyName
                );
              }
            }
            continue;
          }
        }
        const result = new ElementPathResult(
          9 /* hierarchicalToSimple */,
          rawData,
          parentElementPath,
          parentFilePath,
          keyName
        );
        return result;
      }
    } else {
      return new ElementPathResult(0 /* invalid */, null, "", "", null);
    }
  }
  return new ElementPathResult(0 /* invalid */, null, "", "", null);
}

// src/load.ts
var EMPTY_WORKINGDIR_PATH_ERROR = "Error: Cannot load from empty working directory path";
var INVALID_PATH_ERROR = "Error: Invalid path to element on filesystem";
function convertYamlFilePathToElementPath(filePath) {
  if (filePath.slice(-10) === "_this.yaml") {
    const elementPath = filePath.split("/").slice(0, -1).join(".");
    return elementPath;
  } else {
    const elementPath = filePath.slice(0, -5).replace("/", ".");
    return elementPath;
  }
}
function loadYaml(filePath, elementPath, depth) {
  const thisYamlContent = fs2.readFileSync(filePath, "utf8");
  let yamlAsJsObj = yaml2.load(thisYamlContent);
  const keys = Object.keys(yamlAsJsObj);
  if (depth !== 0) {
    if (depth > 0) {
      depth = depth - 1;
    }
    keys.forEach((key) => {
      if (typeof yamlAsJsObj[key] == "string") {
        if (doubleParenthesesRegEx.test(yamlAsJsObj[key])) {
          const dirPath = filePath.split("/").slice(0, -1).join("/");
          const aComplexDataTypeFilePath = trimDoubleParentheses(
            yamlAsJsObj[key]
          );
          if (aComplexDataTypeFilePath.slice(-5) === ".yaml") {
            const newElementPath = convertYamlFilePathToElementPath(
              aComplexDataTypeFilePath
            );
            yamlAsJsObj[key] = load(
              dirPath,
              newElementPath,
              depth
            ).element;
          } else {
            const newFilePathToLoad = path2.join(
              dirPath,
              aComplexDataTypeFilePath
            );
            const aComplexDataTypeContent = fs2.readFileSync(
              newFilePathToLoad,
              "utf-8"
            );
            yamlAsJsObj[key] = aComplexDataTypeContent;
          }
        }
      }
    });
  }
  return new YdsResult(true, yamlAsJsObj, elementPath);
}
function load(workingDirectoryPath, elementPath, depth = -1) {
  if (workingDirectoryPath === "") {
    return new YdsResult(false, null, EMPTY_WORKINGDIR_PATH_ERROR);
  } else {
    let elementPathInfo = getElementPathInfo(workingDirectoryPath, elementPath);
    switch (elementPathInfo.type) {
      case 1 /* empty */:
      case 2 /* shortToObject */:
      case 3 /* shortToList */:
      case 6 /* hierarchicalToObject */:
      case 7 /* hierarchicalToList */:
        return loadYaml(elementPathInfo.data, elementPath, depth);
      case 5 /* shortToSimple */:
      case 9 /* hierarchicalToSimple */:
        return new YdsResult(true, elementPathInfo.data, elementPath);
      case 4 /* shortToComplexString */:
      case 8 /* hierarchicalToComplexString */:
        const elementContent = fs2.readFileSync(elementPathInfo.data, "utf-8");
        return new YdsResult(true, elementContent, elementPath);
      case 0 /* invalid */:
        break;
    }
    return new YdsResult(
      false,
      null,
      INVALID_PATH_ERROR + " [" + workingDirectoryPath + " | " + elementPath + "]"
    );
  }
}

// src/store.ts
import path3 from "path";
import fs3 from "fs";
import yaml3 from "js-yaml";
var INVALID_ELEMENT_NAME = "Error: Invalid element name";
var INVALID_PATH_ERROR2 = "Error: Invalid path";
var NONEMPTY_WORKINGDIR_PATH_ERROR = "Error: Working directory path is non-empty";
var reserved_keywords = [
  "abstract",
  "arguments",
  "async",
  "await",
  "boolean",
  "break",
  "byte",
  "case",
  "catch",
  "char",
  "class",
  "const",
  "continue",
  "debugger",
  "default",
  "delete",
  "do",
  "double",
  "else",
  "enum",
  "eval",
  "export",
  "extends",
  "false",
  "final",
  "finally",
  "float",
  "for",
  "function",
  "goto",
  "if",
  "implements",
  "function",
  "import",
  "in",
  "instanceof",
  "int",
  "interface",
  "let",
  "long",
  "native",
  "new",
  "null",
  "package",
  "private",
  "protected",
  "public",
  "return",
  "short",
  "static",
  "super",
  "switch",
  "synchronized",
  "this",
  "throw",
  "throws",
  "transient",
  "true",
  "try",
  "typeof",
  "using",
  "var",
  "void",
  "volatile",
  "while",
  "with",
  "yield"
];
var idRegex = new RegExp(/^[A-Z0-9]{6}$/);
function validateElementName(elementName) {
  const javascriptVariableNameRegEx = new RegExp(/^[A-Za-z_$][A-Za-z0-9_$]*$/);
  return javascriptVariableNameRegEx.test(elementName) && !reserved_keywords.includes(elementName);
}
function encloseInDoubleParentheses(filePath) {
  return "((" + filePath + "))";
}
function generateComplexStringFilename(elementName, id = "") {
  const containerIsList = id !== "";
  let complexStringFilename = "";
  if (containerIsList) {
    if (elementName.includes("_")) {
      const splitElementName = elementName.split("_");
      if (idRegex.test(splitElementName.slice(-1).toString())) {
        complexStringFilename = elementName + "_" + id;
      } else {
        complexStringFilename = splitElementName.slice(0, -1).join("_") + "_" + id + "." + splitElementName.slice(-1);
      }
    } else {
      complexStringFilename = elementName + "_" + id;
    }
  } else {
    complexStringFilename = complexStringKeyToFileName(elementName);
  }
  return complexStringFilename;
}
function generateObjectOrListFilename(elementName, elementIsArray, id = "") {
  const stringTOAppendToListElement = id === "" ? "" : "_" + id;
  if (elementIsArray) {
    return elementName + stringTOAppendToListElement + ".yaml";
  } else {
    return elementName + stringTOAppendToListElement + "/_this.yaml";
  }
}
function elementNameFromFileName(elementFileName) {
  if (elementFileName.includes("_this.yaml")) {
    return elementFileName.slice(0, -10);
  } else {
    return elementFileName.slice(0, -5);
  }
}
function storeYaml(element, workingDirectoryPath, elementName, depth = -1) {
  let jsObjToSerialize = {};
  let dirPath = "";
  let filename = "";
  const keys = Object.keys(element);
  let container;
  if (Array.isArray(element)) {
    container = 0 /* IsList */;
  } else {
    container = 1 /* IsObject */;
  }
  if (container === 0 /* IsList */) {
    dirPath = workingDirectoryPath;
    filename = elementName + ".yaml";
  } else {
    dirPath = path3.join(workingDirectoryPath, elementName);
    if (depth !== 0 && !fs3.existsSync(dirPath)) {
      fs3.mkdirSync(dirPath);
    }
    filename = "_this.yaml";
  }
  let idCounter = 0;
  keys.forEach((key) => {
    const value = element[key];
    if (typeof value === "string") {
      const stringValue = value;
      if (stringValue.includes("\n")) {
        let complexStringFilename = "";
        if (container === 0 /* IsList */) {
          let id = generateIDs(1, idCounter).pop();
          idCounter++;
          complexStringFilename = generateComplexStringFilename(
            elementName,
            id
          );
        } else {
          complexStringFilename = generateComplexStringFilename(key);
        }
        const complexStringFilePath = path3.join(dirPath, complexStringFilename);
        jsObjToSerialize[key] = encloseInDoubleParentheses(
          complexStringFilename
        );
        if (depth !== 0) {
          fs3.writeFileSync(complexStringFilePath, value);
        }
      } else {
        jsObjToSerialize[key] = value;
      }
    } else if (value === null || typeof value !== "object" || Object.keys(value).length === 0) {
      jsObjToSerialize[key] = value;
    } else {
      let elementFileName = "";
      if (container === 0 /* IsList */) {
        let id = generateIDs(1, idCounter).pop();
        idCounter++;
        elementFileName = generateObjectOrListFilename(
          elementName,
          Array.isArray(value),
          id
        );
      } else {
        elementFileName = generateObjectOrListFilename(
          key,
          Array.isArray(value)
        );
      }
      jsObjToSerialize[key] = encloseInDoubleParentheses(elementFileName);
      if (depth !== 0) {
        if (depth > 0) {
          depth = depth - 1;
        }
        storeYaml(
          value,
          dirPath,
          elementNameFromFileName(elementFileName),
          depth
        );
      }
    }
  });
  if (container === 0 /* IsList */) {
    jsObjToSerialize = Object.values(jsObjToSerialize);
  }
  const filePath = path3.join(dirPath, filename);
  const yamlContentToSerialize = yaml3.dump(jsObjToSerialize);
  if (idCounter > 0) {
    const listMetadataFilePath = path3.join(dirPath, "." + filename);
    const listMetadata = { idCounter };
    const listMetadataAsYaml = yaml3.dump(listMetadata);
    fs3.writeFileSync(listMetadataFilePath, listMetadataAsYaml, "utf-8");
  }
  fs3.writeFileSync(filePath, yamlContentToSerialize, "utf-8");
  return new YdsResult(true, element, elementName);
}
function store(element, workingDirectoryPath, elementName) {
  if (fs3.existsSync(workingDirectoryPath)) {
    if (fs3.readdirSync(workingDirectoryPath).length > 0) {
      return new YdsResult(
        false,
        element,
        NONEMPTY_WORKINGDIR_PATH_ERROR + " [" + workingDirectoryPath + "]"
      );
    } else {
      if (validateElementName(elementName)) {
        return storeYaml(element, workingDirectoryPath, elementName);
      } else {
        return new YdsResult(
          false,
          element,
          INVALID_ELEMENT_NAME + " [" + elementName + "]"
        );
      }
    }
  }
  return new YdsResult(
    false,
    element,
    INVALID_PATH_ERROR2 + " [" + workingDirectoryPath + "]"
  );
}

// src/delete.ts
import path4 from "path";
import fs4 from "fs";
import yaml4 from "js-yaml";
var DELETE_EMPTY_ELEMENT_PATH_ERROR = "Error: Cannot delete element for empty element path";
function recursivelyDeleteList(listPath) {
  const listFilePath = path4.join(listPath.dir, listPath.base);
  const listContents = fs4.readFileSync(listFilePath, "utf-8");
  const listAsJsObj = yaml4.load(listContents);
  for (const listItem of listAsJsObj) {
    if (doubleParenthesesRegEx.test(listItem)) {
      const listItemPath = path4.parse(
        path4.join(listPath.dir, trimDoubleParentheses(listItem))
      );
      const listItemFilePath = path4.join(listItemPath.dir, listItemPath.base);
      if (listItemPath.base === "_this.yaml") {
        fs4.rmSync(listItemPath.dir, { recursive: true });
      } else if (listItemPath.ext === ".yaml") {
        recursivelyDeleteList(listItemPath);
      } else {
        fs4.rmSync(listItemFilePath);
      }
    }
  }
  fs4.rmSync(listFilePath);
}
function deleteChildFromParentElement(parentElement, childElement) {
  if (Array.isArray(parentElement)) {
    parentElement.splice(childElement, 1);
  } else {
    delete parentElement[childElement];
  }
}
function deleteElement(workingDirectoryPath, elementPath, depth = 0) {
  if (workingDirectoryPath === "") {
    return new YdsResult(false, null, EMPTY_WORKINGDIR_PATH_ERROR);
  } else {
    const elementPathInfo = getElementPathInfo(
      workingDirectoryPath,
      elementPath
    );
    const parentElementPath = elementPathInfo.parentElementPath;
    const parentFilePath = elementPathInfo.parentFilePath;
    const parentIsAnElement = elementPathInfo.parentIsElement;
    let parentElement;
    if (parentIsAnElement && fs4.existsSync(parentFilePath)) {
      const parentElementFileContents = fs4.readFileSync(
        parentFilePath,
        "utf-8"
      );
      parentElement = yaml4.load(parentElementFileContents);
    }
    switch (elementPathInfo.type) {
      case 1 /* empty */:
        return new YdsResult(
          false,
          null,
          DELETE_EMPTY_ELEMENT_PATH_ERROR + " [" + workingDirectoryPath + " | " + elementPath + "]"
        );
      case 2 /* shortToObject */:
      case 6 /* hierarchicalToObject */:
        const objectFilePath = path4.parse(elementPathInfo.data).dir;
        fs4.rmSync(objectFilePath, { recursive: true });
        if (parentIsAnElement && fs4.existsSync(parentFilePath)) {
          deleteChildFromParentElement(parentElement, elementPathInfo.keyName);
          const parentElementOfObjectContentsToStore = yaml4.dump(parentElement);
          fs4.writeFileSync(
            parentFilePath,
            parentElementOfObjectContentsToStore
          );
          const parentElementOfObjectStoredToDisk = load(
            workingDirectoryPath,
            parentElementPath,
            depth
          ).element;
          return new YdsResult(
            true,
            parentElementOfObjectStoredToDisk,
            parentElementPath
          );
        }
        return new YdsResult(true, null, "");
      case 3 /* shortToList */:
      case 7 /* hierarchicalToList */:
        const listFilePath = elementPathInfo.data;
        const listPath = path4.parse(listFilePath);
        const listParentDir = listPath.dir;
        recursivelyDeleteList(listPath);
        const listMetadataFilePath = path4.join(
          listPath.dir,
          "." + listPath.base
        );
        const directoryContents = fs4.readdirSync(listParentDir);
        if (directoryContents.includes("." + listPath.base)) {
          fs4.rmSync(listMetadataFilePath);
        }
        deleteChildFromParentElement(parentElement, elementPathInfo.keyName);
        const parentElementOfListContentsToStore = yaml4.dump(parentElement);
        fs4.writeFileSync(
          parentFilePath,
          parentElementOfListContentsToStore,
          "utf-8"
        );
        const parentElementOfListStoredToDisk = load(
          workingDirectoryPath,
          parentElementPath,
          depth
        ).element;
        return new YdsResult(
          true,
          parentElementOfListStoredToDisk,
          parentElementPath
        );
      case 5 /* shortToSimple */:
      case 9 /* hierarchicalToSimple */:
        deleteChildFromParentElement(parentElement, elementPathInfo.keyName);
        fs4.writeFileSync(parentFilePath, yaml4.dump(parentElement));
        return new YdsResult(true, parentElement, parentElementPath);
      case 4 /* shortToComplexString */:
      case 8 /* hierarchicalToComplexString */:
        const complexStringFilePath = elementPathInfo.data;
        fs4.rmSync(complexStringFilePath);
        deleteChildFromParentElement(parentElement, elementPathInfo.keyName);
        const parentElementOfComplexStringContentsToStore = yaml4.dump(parentElement);
        fs4.writeFileSync(
          parentFilePath,
          parentElementOfComplexStringContentsToStore,
          "utf-8"
        );
        const parentElementOfComplexStringStoredToDisk = load(
          workingDirectoryPath,
          parentElementPath,
          depth
        ).element;
        return new YdsResult(
          true,
          parentElementOfComplexStringStoredToDisk,
          parentElementPath
        );
      case 0 /* invalid */:
        break;
    }
    return new YdsResult(
      false,
      null,
      INVALID_PATH_ERROR + " [" + workingDirectoryPath + " | " + elementPath + "]"
    );
  }
}

// src/clear.ts
import path5 from "path";
import fs5 from "fs";
import yaml5 from "js-yaml";
var CLEAR_EMPTY_ELEMENT_PATH_ERROR = "Error: Cannot clear element for empty element path";
function clear(workingDirectoryPath, elementPath, depth = 0) {
  if (workingDirectoryPath === "") {
    return new YdsResult(false, null, EMPTY_WORKINGDIR_PATH_ERROR);
  } else {
    const elementPathInfo = getElementPathInfo(
      workingDirectoryPath,
      elementPath
    );
    const parentElementPath = elementPathInfo.parentElementPath;
    const parentFilePath = elementPathInfo.parentFilePath;
    const parentIsAnElement = elementPathInfo.parentIsElement;
    let parentElement;
    if (parentIsAnElement && fs5.existsSync(parentFilePath)) {
      const parentElementFileContents = fs5.readFileSync(
        parentFilePath,
        "utf-8"
      );
      parentElement = yaml5.load(parentElementFileContents);
    }
    switch (elementPathInfo.type) {
      case 1 /* empty */:
        return new YdsResult(
          false,
          null,
          CLEAR_EMPTY_ELEMENT_PATH_ERROR + " [" + workingDirectoryPath + " | " + elementPath + "]"
        );
      case 2 /* shortToObject */:
        const shortToObjectParsedPath = path5.parse(elementPathInfo.data);
        const shortToObjectFilePath = shortToObjectParsedPath.dir;
        if (!parentIsAnElement) {
          fs5.rmSync(shortToObjectFilePath, { recursive: true });
          fs5.mkdirSync(shortToObjectFilePath);
          const elementToStore = {};
          const objectContentsToStore = yaml5.dump(elementToStore);
          const thisYamlFilePath = path5.join(
            shortToObjectFilePath,
            "_this.yaml"
          );
          fs5.writeFileSync(thisYamlFilePath, objectContentsToStore);
          return new YdsResult(true, elementToStore, elementPathInfo.keyName);
        }
      case 6 /* hierarchicalToObject */:
        const objectParsedPath = path5.parse(elementPathInfo.data);
        const objectFilePath = objectParsedPath.dir;
        fs5.rmSync(objectFilePath, { recursive: true });
        if (parentIsAnElement && fs5.existsSync(parentFilePath)) {
          parentElement[elementPathInfo.keyName] = {};
          const parentElementOfObjectContentsToStore = yaml5.dump(parentElement);
          fs5.writeFileSync(
            parentFilePath,
            parentElementOfObjectContentsToStore
          );
        }
        return new YdsResult(true, parentElement, parentElementPath);
      case 3 /* shortToList */:
      case 7 /* hierarchicalToList */:
        const listFilePath = elementPathInfo.data;
        const listPath = path5.parse(listFilePath);
        const listParentDir = listPath.dir;
        recursivelyDeleteList(listPath);
        const listMetadataFilePath = path5.join(
          listPath.dir,
          "." + listPath.base
        );
        const directoryContents = fs5.readdirSync(listParentDir);
        if (directoryContents.includes("." + listPath.base)) {
          fs5.rmSync(listMetadataFilePath);
        }
        parentElement[elementPathInfo.keyName] = [];
        const parentElementOfListContentsToStore = yaml5.dump(parentElement);
        fs5.writeFileSync(
          parentFilePath,
          parentElementOfListContentsToStore,
          "utf-8"
        );
        const parentElementOfListStoredToDisk = load(
          workingDirectoryPath,
          parentElementPath,
          depth
        ).element;
        return new YdsResult(
          true,
          parentElementOfListStoredToDisk,
          parentElementPath
        );
      case 5 /* shortToSimple */:
      case 9 /* hierarchicalToSimple */:
        if (parentElement[elementPathInfo.keyName] === null || parentElement[elementPathInfo.keyName] === "" || typeof parentElement[elementPathInfo.keyName] === "object") {
          return new YdsResult(true, parentElement, parentElementPath);
        } else if (typeof parentElement[elementPathInfo.keyName] === "string") {
          parentElement[elementPathInfo.keyName] = "";
          fs5.writeFileSync(parentFilePath, yaml5.dump(parentElement));
          return new YdsResult(true, parentElement, parentElementPath);
        } else if (typeof parentElement[elementPathInfo.keyName] === "number" || typeof parentElement[elementPathInfo.keyName] === "boolean") {
          parentElement[elementPathInfo.keyName] = null;
          fs5.writeFileSync(parentFilePath, yaml5.dump(parentElement));
          return new YdsResult(true, parentElement, parentElementPath);
        }
      case 4 /* shortToComplexString */:
      case 8 /* hierarchicalToComplexString */:
        const complexStringFilePath = elementPathInfo.data;
        const complexStringContentsToStore = "\n";
        fs5.writeFileSync(
          complexStringFilePath,
          complexStringContentsToStore,
          "utf-8"
        );
        return new YdsResult(true, parentElement, parentElementPath);
      case 0 /* invalid */:
        break;
    }
    return new YdsResult(
      false,
      null,
      INVALID_PATH_ERROR + " [" + workingDirectoryPath + " | " + elementPath + "]"
    );
  }
}

// src/result.ts
var YdsResult = class {
  _success;
  _element;
  _message;
  /**
   * Default constructor for YdsResult
   *
   * @param success success status of a yaml-datastore operation
   * @param element element read into memory or stored on-disk per yaml-datastore operation
   * @param message message describing success status of a yaml-datastore operation
   * @returns new YdsResult object
   */
  constructor(success, element, message) {
    this._success = success;
    if (this._success) {
      this._element = element;
    } else {
      this._element = null;
    }
    this._message = message;
  }
  /** @returns success status of a yaml-datastore operation. */
  get success() {
    return this._success;
  }
  /** @returns element read into memory or stored on-disk per yaml-datastore operation. */
  get element() {
    return this._element;
  }
  /** @returns message describing success status of a yaml-datastore operation. */
  get message() {
    return this._message;
  }
};
export {
  YdsResult,
  clear,
  deleteElement,
  generateIDs,
  load,
  store
};
